File: client/src/libs/api/models/article.ts
import { Tag } from "./tag";
import { Topic } from "./topic";

export enum ArticleOrderBy {
  PUBLISHED_AT_DESC = "published_at_desc",
  PUBLISHED_AT_ASC = "published_at_asc",
}

export type ArticleQueryParams = {
  orderBy?: ArticleOrderBy;
  keyword?: string;
};

export type Article = {
  articleId: string;
  title: string;
  content: string;
  publishedAt?: string;
  tags: Tag[];
  topic?: Topic;
};

File: go_api/presentation/serializer/article.go
package serializer

import (
	"time"

	"github.com/google/uuid"
	"github.com/nozomi-iida/nozo_blog/domain/article"
)

type ArticleJson struct {
	ArticleID   uuid.UUID  `json:"articleId"`
	Title       string     `json:"title"`
	Content     string     `json:"content"`
	PublishedAt *time.Time `json:"publishedAt,omitempty"`
	Tags        []TagJson  `json:"tags"`
	Topic       *TopicJson `json:"topic,omitempty"`
	Author      UserJson   `json:"author"`
}

func Article2Json(article article.ArticleDto) ArticleJson {
	tags := []TagJson{}
	if len(article.Tags) > 0 {
		for _, tag := range article.Tags {
			tags = append(tags, Tag2Json(tag))
		}
	}
	var topic TopicJson
	if article.Topic != nil {
		topic = Topic2Json(*article.Topic)
	}

	return ArticleJson{
		ArticleID:   article.ArticleID,
		Title:       article.Title,
		Content:     article.Content,
		PublishedAt: article.PublishedAt,
		Tags:        tags,
		Topic:       &topic,
		Author:      User2Json(article.Author),
	}
}
